subsys/usb/class/bluetooth.c

```diff
@@ -167,32 +167,90 @@ static void hci_rx_thread(void)
 	}
 }
 
+static uint16_t hci_pkt_get_len(struct net_buf *buf,
+				const uint8_t *data, size_t size)
+{
+	uint16_t len = 0;
+	size_t hdr_len = 0;
+
+	switch (bt_buf_get_type(buf)) {
+	case BT_BUF_CMD: {
+		struct bt_hci_cmd_hdr *cmd_hdr;
+
+		hdr_len = sizeof(*cmd_hdr);
+		cmd_hdr = (struct bt_hci_cmd_hdr *)data;
```

Comment:
This is style widely followed in Zephyr source code.

Reply:
I would disagree, it is widely used in Bluetooth subsystem.

Reply:
ok, then there is inconsistency in USB subsystem

![image](https://user-images.githubusercontent.com/6350656/107238099-be113680-6a27-11eb-9297-87ea8122a897.png)

Reply:
well, I think in this case `git grep "= (void \*)"` is more correct. And I am not talking about inconsistency, just that it is widely used in Bluetooth (541 + 8 + 69 vs 189 anywhere else). And my question was: _What is advantage here to cast to (void *)?_ not where it is used how much.

Reply:
@jhedberg @pabigot any comments on this?

Reply:
I've always thought that the main reason for using `(void *)` typecasts is that it's almost always shorter than spelling out the full type :)

Reply:
> I've always thought that the main reason for using `(void *)` typecasts is that it's almost always shorter than spelling out the full type :)

Yes that is what I am thinking too. But is it relevant to MISRA Rule 11.5 "A conversion should not be performed from pointer to void into pointer to object" (we do not follow this rule, but still interesting)? Is it always okay to cast probably unaligned buffer to pointer to void, ant than into pointer to object?

Reply:
> I've always thought that the main reason for using `(void *)` typecasts is that it's almost always shorter than spelling out the full type :)

That's probably the only viable justification here.  I'd rather see the function parameter be `const void *`, but that's a major refactoring.

> Is it always okay to cast probably unaligned buffer to pointer to void, and than into pointer to object?

No.  The cast to `void *` discards alignment information, so you can break things using that idiom that the compiler might be able to diagnose otherwise.

The relevant part (from C17) is 6.3.2.3.7:

> A pointer to an object type may be converted to a pointer to a different object type. If the resulting pointer is not correctly aligned for the referenced type, the behavior is undefined. Otherwise, when converted back again, the result shall compare equal to the original pointer.

Here `struct bt_hci_cmd_hdr` is `__packed` so it seems OK.

IMO leave it as:

    cmd_hdr = (struct bt_hci_cmd_hdr *)data;

The use of the `(void *)` cast idiom in the USB subsystem is just weird, and based on a quick review probably derives from the personal style from an original author back in 2017, weighted by the then-"enforced" 80 character line limit.

I suspect a coccinelle rule would fix that right up....

Reply:
We follow the Linux kernel coding style, and found this under "Allocating Memory"

"Casting the return value which is a void pointer is redundant. The conversion from void pointer to any other pointer type is guaranteed by the C programming language."

Here, I know `data` is not void pointer, neither it is guaranteed to be word aligned due to it being `uint8_t *` which could have as well have been a `void *` type in the signature of the function. We are anyway converting a pointer into to a pointer to object as an intended implementation requirement to decode the HCI packet.

`void *` is definitely shorter to type, but also is better at readability and maintainability in cases where in the future the pointer variable type is changed.

That said, I have no strong argument against keeping the code in this PR as-is

Reply:
> `void *` is definitely shorter to type, but also is better at readability and maintainability in cases where in the future the pointer variable type is changed.

I disagree fairly strongly with that statement.  It's never an improvement to maintainability to throw away type information in pointer assignment.  If one of the types was changed to an unpacked structure, code would break, and C would allow it since it assumes what it's told to do is correct.

But it's no longer relevant to this PR.

Reply:
You are ignoring `struct bt_hci_cmd_hdr *cmd_hdr;`, C will do the right thing, it knows `cmd_hdr` is `__packed`.

Reply:
> If one of the types was changed to an unpacked structure

Reply:
> 
> 
> > If one of the types was changed to an unpacked structure

`cmd_hdr = (struct bt_hci_cmd_hdr *)data;`  does not help in that case, C will do the right thing and CPU will fail accessing unaligned memory.